core/app/models/spree/promotion/rules/user_role.rb

```diff
@@ -0,0 +1,26 @@
+module Spree
+  class Promotion
+    module Rules
+      class UserRole < PromotionRule
+        has_many :promotion_rule_roles,
+          class_name: 'Spree::PromotionRuleRole',
+          foreign_key: :promotion_rule_id,
+          dependent: :destroy
+
+        has_many :roles,
+          through: :promotion_rule_roles,
+          class_name: 'Spree::Role'
+
+        accepts_nested_attributes_for :roles
+
+        def applicable?(promotable)
+          promotable.is_a?(Spree::Order)
+        end
+
+        def eligible?(order, options = {})
+          (roles & order.user.spree_roles).any? if order.user
```

Comment:
I'm a toss up on this. When we first discussed it In Real Life at solidusconf I was a fan of it going into core, but I didn't realize it would need another table with it. I see arguments to leave it out and make it a public extension or bring it in as its an incredibly common request.

I'm less swayed by the arguments of it being potentially abused. I'd rather burn that bridge when we come to it.

@jordan-brough @gmacdougall @tvdeyen opinions for bringing it in or leaving it out of core?

Good code though, I'll definitely end up using this either way.